File: source/plate_recognizer.cpp
#include <iostream>

#include "plate_recognizer.h"   

using namespace std;


PlateRecognizer::PlateRecognizer() 
{
    net = cv::dnn::readNet("models/plate_recognizer.onnx");
    net.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);
    net.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA);

    mean = cv::Scalar(123.68, 116.78, 103.94);
    plate_width = 320;
    plate_height = 64;
    scalefactor = 1.0;
    score_threshold = 0.90;
    alphabet = "##0123456789ABCDEFGHIJKLMNPQRSTUVWXYZ";
}

PlateRecognizer::~PlateRecognizer()
{
    
}

bool PlateRecognizer::recognize(Plate& plate)
{
    cv::Mat test;
    start_time = clock();
    
    cv::dnn::blobFromImage(plate.image, blob, scalefactor, cv::Size(plate_width, plate_height), mean, true, false); // Create a 4D blob from a plate image.
    
    net.setInput(blob);
    net.forward(outs);

    cv::Mat input = outs.reshape(1, vector<int> {outs.size[0], outs.size[2]});
    plate.text = postProcess(input);

    end_time = clock();
    cout << "recognize forward time: " << float(end_time - start_time) / CLOCKS_PER_SEC << " s" << endl;

    return plate.text->size() > 0;
}

string PlateRecognizer::postProcess(cv::Mat input)
{
    cv::Mat softmax_result, values, probs, final_probs;
    double minVal, maxVal;
    cv::Point minLoc, maxLoc;
    cv::Scalar mean;
    string text = "";

    softmax(input, softmax_result);
    for (int i = 0; i < softmax_result.size[0]; i++)
    {
        cv::minMaxLoc(softmax_result.row(i), &minVal, &maxVal, &minLoc, &maxLoc);
        probs.push_back(maxVal);
        values.push_back(maxLoc.x);
    }
//    cout << "softmax_result" << softmax_result << endl;

    for (int i = 0; i < values.size[0]; i++)
    {
        cout << values.at<int>(i) << endl;
        if(values.at<int>(i) != 0 && !(i > 0 && values.at<int>(i) == values.at<int>(i-1)))
        {
            text += alphabet[values.at<int>(i)];
            final_probs.push_back(probs.at<double>(i));
        }
    }
    mean = cv::mean(final_probs);
    cout << "score " << mean[0] << endl;

    if(mean[0] > score_threshold)
        return text;
    else
        return "";
}

void softmax(cv::Mat input, cv::Mat& output)
{
    cv::Mat e_x;
    double minVal, maxVal;
    cv::Point minLoc, maxLoc;

    for (int i = 0; i < input.size[0]; i++)
    {
        cv::minMaxLoc(input.row(i), &minVal, &maxVal, &minLoc, &maxLoc);
        cv::exp(input.row(i) - maxVal, e_x);
        output.push_back(e_x / cv::sum(e_x));
    }
}

File: python/source/plate_recognizer.py
import re
from .functions import *
from .plate import Plate


class PlateRecognizer:
    alphabet = "#0123456789ABCDEFGHIJKLMNPQRSTUVWXYZ"

    def __init__(self):
        self.net = cv2.dnn.readNet("models/plate_recognizer.onnx")
        self.net.setPreferableBackend(cv2.dnn.DNN_BACKEND_CUDA)
        self.net.setPreferableTarget(cv2.dnn.DNN_TARGET_CUDA)

        self.mean = (123.68, 116.78, 103.94)
        self.scale_factor = 1.0
        self.threshold = 0.9

    def recognize(self, plate):
        try:
            image = cv2.cvtColor(plate.image, cv2.COLOR_RGB2BGR)

            blob = cv2.dnn.blobFromImage(image, size=(320, 64), mean=self.mean, scalefactor=self.scale_factor)
            self.net.setInput(blob)

            # Run the recognition model
            result = self.net.forward()

            softmax_result = self.softmax(result)
            values = np.max(softmax_result, axis=2).flatten()
            prob = np.argmax(softmax_result, axis=2).flatten()

            preds_idx = np.nonzero(prob)
            confidence = np.mean(values[preds_idx])

            # decode the result into text
            text = self.decodeText(result)

            # find char index in text
            find_index = re.search(r"[a-zA-Z]", text, re.I)
            if find_index is not None:
                i = find_index.start()
                if len(text) >= 8 and confidence >= self.threshold:
                    plate.text = text[i-2:i] + ' ' + alphabetDict(text[i]) + ' ' + text[i+1:i+4] + ' - ' + text[i+4:i+6]
                    print(f'OCR process done for plate id {plate.id}, confidence: {confidence}')

        except Exception as e:
            print(f'error in recognize function:', e)

    def decodeText(self, scores):
        text = ""
        for i in range(scores.shape[0]):
            c = np.argmax(scores[i][0])
            if c != 0:
                text += self.alphabet[c - 1]
            else:
                text += '-'

        # adjacent same letters as well as background text must be removed to get the final output
        char_list = []
        for i in range(len(text)):
            # print(text)
            if text[i] != '-' and not (i > 0 and text[i] == text[i - 1]):
                char_list.append(text[i])

        return ''.join(char_list)

    def softmax(self, a):
        """Compute softmax values for each sets of scores in x."""
        result = []
        for x in a:
            e_x = np.exp(x - np.max(x))
            result.append(e_x / e_x.sum(axis=1))

        return np.array(result)
